### ISSUE TITLE
lint to disallow storage of internationalized strings to static finals

### ISSUE THREAD

**Issue by pq** [dart-lang/linter#2085]
To enable runtime language selection, results  from calling `Intl.message` should not be stored in static final, or objects that hold them.

**Comment by bwilkerson** [comment]
> ... or objects that hold them.

I don't know what that means. Could you provide an example?

**Comment by pq** [comment]
Sorry, yes.  I should have said "details to follow" after some research.  This is just a placeholder for now.

**Comment by jefflim-google** [comment]
As an example of 'objects that hold them', if you have:
```
class Country {   
  String countryName;
  Country(this.countryName);
}
```
and then:
```
  static final singapore = Country(strings.singapore);
```
Then the string inside country could never change, even if the user changes language inside the app.

**Comment by pq** [comment]
For more context, `strings` as referenced above is a holder for String constants that are built up (@jefflim-google correct me if I'm wrong) with calls to `Intl.message` that look like this:


```dart
/// strings.dart

String get myMessage => Intl.message(
      'Hi',
      name: 'myMessage',
      desc: 'A message...',
    );
```

**Comment by bwilkerson** [comment]
Thanks for the context!

At the point at which the linter sees `strings.singapore` it will only have the element for the getter `singapore` and it won't be able to tell that it's returning a value computed by invoking `Intl.message`. It would be too expensive to get the AST for every library that define a getter that returns a `String`, so we'll need some other way to identify invocations of getters that return localized strings. There might be enough information in the element model (if `strings`, for example, is always a prefix for a library that imports from `package:intl`).

If we can identify the getter invocations that return localized strings, and if those getters are always either the initializers of static fields or the arguments to constructor invocations,, then we should be able to do a reasonable job of finding problems.

A full solution, however, would require full data flow analysis, which would likely be too expensive for a lint. Imagine, for example, the `Country` class given above and a utility method like the following:
```
Country buildCountry(String countryCode) => Country(countryFromCode(countryCode));

String countryFromCode(String countryCode) {
  if (countryCode == 'sg') {
    return strings.singapore;
  }
}
```
To be completely accurate, we'd need to find all references to the getter `strings.singapore`, see that the value is being returned from `countryFromCode`, final all references to `countryFromCode`, etc.

**Comment by jefflim-google** [comment]
We've been consistently naming these with a 'strings' import, i.e. we always import these as:

import 'strings.dart' as strings'

And we always invoke them as: strings.xxxx or strings.xxxx().

We'd be happy with a solution that is just based on this, to avoid the analysis required.

We'd also be happy to forego full data flow analysis with something that catches 90% of the simple cases, even if it misses some deeper issues.

**Comment by pq** [comment]
Thanks @jefflim-google!

My preference would be to generalize this as much as possible.  To that end, @bwilkerson and I were chatting about how annotations might help here and we came up with an idea that I'd like to float by you.

**Proposal:** a new annotation`@doNotStore` to tag values that should not be stored in fields (or be returned from functions/getters that are not similarly tagged).

The annotation could be applied at the class, class member or library level.

In the case of a `strings.dart` file it would look something like:

```dart
@doNotStore
library strings;

import 'package:intl/intl.dart';

String get message1 => Intl.message(
      'Message 1',
      name: 'message1',
      desc: 'Default text for message 1 [CHAR_LIMIT=12]',
    );

String get message2 => Intl.message(
      'Message 2',
      name: 'message2',
      desc: 'Default text for message 2 [CHAR_LIMIT=12]',
    );

...
```

**Comment by jefflim-google** [comment]
@doNotStore sounds like an excellent approach to me!

**Comment by jefflim-google** [comment]
One subtlety: It's fine to store a *function* into a static final or const map:

String message1() => Intl.message(
      'Message 1',
      name: 'message1',
      desc: 'Default text for message 1 [CHAR_LIMIT=12]',
    );

const lookup = <String, String Function()>{ key1, message1 };

Will this support that?

**Comment by pq** [comment]
Restricting the lint to  flag invocations (and allow references) seems fine to me.

In case you are holding onto function references, you'll want to annotate `lookup` itself as well.

@bwilkerson : do you see  any wrinkles I'm missing?

**Comment by bwilkerson** [comment]
Nope. Analyzer knows the difference between a reference and an invocation, so it should be easy to allow references to functions.
